src/OpenFeature/OpenFeatureClient.cs

```diff
@@ -9,69 +12,171 @@ namespace OpenFeature
-            GetObjectDetails(flagKey, defaultValue, context, config).Value;
+        public async Task<T> GetObjectValue<T>(string flagKey, T defaultValue, EvaluationContext context = null, FlagEvaluationOptions config = null)  =>
+            (await GetObjectDetails(flagKey, defaultValue, context, config)).Value;
 
-        public FlagEvaluationDetails<T> GetObjectDetails<T>(string flagKey, T defaultValue, EvaluationContext context = null, FlagEvaluationOptions config = null)  =>
-            EvaluateFlag(OpenFeature.GetProvider().ResolveStructureValue, flagKey, defaultValue, context, config);
-        
-        private delegate ResolutionDetails<T> ResolveValueDelegate<T>(string flagKey, T defaultValue, EvaluationContext context, FlagEvaluationOptions config);
-        
-        private static FlagEvaluationDetails<T> EvaluateFlag<T>(ResolveValueDelegate<T> resolveValueDelegate, string flagKey, T defaultValue, EvaluationContext context = null, FlagEvaluationOptions config = null)
+        public async Task<FlagEvaluationDetails<T>> GetObjectDetails<T>(string flagKey, T defaultValue, EvaluationContext context = null, FlagEvaluationOptions config = null)  =>
+            await EvaluateFlag(_featureProvider.ResolveStructureValue, FlagValueType.Object, flagKey, defaultValue, context, config);
+
+        private async Task<FlagEvaluationDetails<T>> EvaluateFlag<T>(
+            Func<string, T, EvaluationContext, FlagEvaluationOptions, Task<ResolutionDetails<T>>> resolveValueDelegate,
+            FlagValueType flagValueType, string flagKey, T defaultValue, EvaluationContext context = null, FlagEvaluationOptions options = null)
         {
             // New up a evaluation context if one was not provided.
             if (context == null) 
             {
                 context = new EvaluationContext();
             }
-            
+
+            var evaluationContext = OpenFeature.GetContext();
+            evaluationContext.Merge(context);
+
+            var allHooks = new List<IHook>()
+                .Concat(OpenFeature.GetHooks())
+                .Concat(_hooks)
+                .ToList()
+                .AsReadOnly();
+
+            var allHooksReversed = allHooks
+                .AsEnumerable()
+                .Reverse()
+                .ToList()
+                .AsReadOnly();
+
+            var hookContext = new HookContext<T>(
+                flagKey,
+                defaultValue,
+                flagValueType,
+                _metadata,
+                OpenFeature.GetProviderMetadata(),
+                evaluationContext
+            );
+
             FlagEvaluationDetails<T> response;
             try
             {
-                response = resolveValueDelegate.Invoke(flagKey, defaultValue, context, config)
+                await TriggerBeforeHooks(allHooks, hookContext, options);
+
+                response = (await resolveValueDelegate.Invoke(flagKey, defaultValue, hookContext.EvaluationContext, options))

```

Comment:
Total nitpick, but I'm not a huge fan of `response`. `resolution`, `evaluation` or something might be better.

Reply:
Done 👍🏻